src/collider/BVHTree.cpp

```diff
@@ -158,3 +160,252 @@ BVHNode *BVHTree::AllocNode()
 		Log::Error("Out of space in m_bvhNodes.");
 	return &m_bvhNodes[m_nodeAllocPos++];
 }
+
+double BVHTree::CalculateSAH() const
+{
+	double outSAH = 0.0;
+	std::vector<BVHNode *> m_nodeStack = { m_root };
+
+	while (!m_nodeStack.empty()) {
+		BVHNode *node = m_nodeStack.back();
+		m_nodeStack.pop_back();
+
+		if (!node->IsLeaf() && node->kids[0])
+			m_nodeStack.push_back(node->kids[0]);
+		if (!node->IsLeaf() && node->kids[1])
+			m_nodeStack.push_back(node->kids[1]);
+
+		// surface area = 2 * width * depth * height
+		vector3d size = node->aabb.max - node->aabb.min;
+		double area = 2.f * size.x * size.y * size.z;
+
+		// Cost function according to https://users.aalto.fi/~laines9/publications/aila2013hpg_paper.pdf Eq. 1
+		if (node->IsLeaf())
+			outSAH += node->numTris * area;
+		else
+			outSAH += 1.2 * area;
+	}
+
+	vector3d rootSize = m_root->aabb.max - m_root->aabb.min;
+	float rootArea = 2.f * rootSize.x * rootSize.y * rootSize.z;
```

Comment:
This calculates a float but does all the calculations along the way in double precision bar the final multiplication by `2.f` (I think).  Is this intended?  I no-longer know on modern hardware what the cost is vs the cost of say casting the min/max bb to float and then doing it.  If you did that you could also cast them to SIMD aligned float vectors for some speed increases, might even find that that if our SIMD float vector 3 has an implicit `w` of 1.0f always the last operation here can be 4 component dot product and even faster.

Reply:
The surface-area-heuristic calculation function was only implemented for debugging purposes (I used it to validate the correctness of the replacement BVHTree implementation) and was banged together quite quickly. There's no rhyme or reason behind the datatype usage here, other than that it was originally written with `float` and the return value of the function changed afterwards. Will at the least make it consistent and use `double` the whole time.